### PR TITLE
ALiBi Implementation

### PR DESCRIPTION
**Pull request by ofirpress**
I've implemented ALiBi efficiently, by just modifying the matmul_result matrix in transformer.py 

On my hardware, with the pretrain_gpt.py example, ALiBi runs just as fast as sinusoidal and uses 20MB of extra memory (that is 0.2% of the total memory usage). 

I think this implementation could be made even more efficient by modifying the mask instead of matmul_result, but this would require modifying the cuda code in scaled_upper_triang_masked_softmax and I don't have that knowledge.

### CODE REVIEW

File: megatron/model/transformer.py [hunk 1]
```diff
@@ -203,7 +203,7 @@ def __init__(self, init_method,
             self.rotary_emb = RotaryEmbedding(self.hidden_size_per_attention_head, precision=args.params_dtype)
 
     def forward(self, hidden_states, attention_mask, layer_past=None,
-                get_key_value=False, encoder_output=None):
+                get_key_value=False, encoder_output=None, alibi=None):
```
Comment (ibeltagy) [inline megatron/model/transformer.py:206]: nit: probably nicer to pass this arg using `self.position_embedding_type` as in  `if self.position_embedding_type == PositionEmbeddingType.rotary` and avoid passing `alibi=None` in all the function calls.

File: megatron/model/transformer.py [hunk 2]
```diff
@@ -660,11 +684,18 @@ def build_layer(layer_number):
             get_cuda_rng_tracker = deepspeed.checkpointing.get_cuda_rng_tracker
             checkpoint = deepspeed.checkpointing.checkpoint
 
+        if args.position_embedding_type == PositionEmbeddingType.alibi:
+            self.alibi = self._build_alibi_tensor(args.seq_length, args.num_attention_heads, args.micro_batch_size).to(torch.cuda.current_device())
```
Comment (ibeltagy) [inline megatron/model/transformer.py:688]: Is `.to(torch.cuda.current_device())` safe in the multi-gpu model parallelism setup? 
@stas00, @slippylolo do you know?

File: megatron/model/transformer.py [hunk 3]
```diff
@@ -660,11 +684,18 @@ def build_layer(layer_number):
...
+            if args.params_dtype == torch.float16:
+                self.alibi = self.alibi.half()
```
Comment (ibeltagy) [inline megatron/model/transformer.py:690]: Is this the right way to convert to fp16? how about bf16 (check `args.bf16`)?
@stas00 might know?

**Review by ibeltagy — APPROVED** [review]
Looks great, thanks @ofirpress. 

I left a couple of questions, mainly wondering if certain constructs are safe to use in the model parallelism setup.

File: megatron/model/transformer.py (hunk 1 above)
Comment (ofirpress) [reply megatron/model/transformer.py:206]: I'm not sure what you mean here? alibi here is an object that has the alibi matrix, it's not just a boolean. Thanks

File: megatron/model/transformer.py (hunk 2 above)
Comment (stas00) [reply megatron/model/transformer.py:688]: This needs to be threaded with care. But on the other hand things quickly break if it's not done right and pytorch let's you know quickly ;)

It should already allocate the tensor on the current device in `__init__()`. does it break if don't `to()` explicity, @ofirpress?

Note that `matmul_result` is created during `forward` which indeed has the correct device set.

The other approach to not needing to use `torch.cuda.current_device()` in forward is to use `.device` of one of the inputs.

Comment (stas00) [reply megatron/model/transformer.py:688]: They key is to add a good test and we have a multi-gpu CI, so it should be easy to validate.

File: megatron/model/transformer.py (hunk 3 above)
Comment (stas00) [reply megatron/model/transformer.py:690]: yes, that's correct. it's just a synonym to `.to(torch.float16)`

and yes, it's probably a good idea to plan for bf16 and do: `.to(torch.bfloat16)`

**Comment by stas00** [comment]
> I think this implementation could be made even more efficient by modifying the mask instead of matmul_result, but this would require modifying the cuda code in scaled_upper_triang_masked_softmax and I don't have that knowledge.

It's always good to start simple and then improve. So you could complete this work using your expertise, write good tests and make it a solid component and merge it. Then post an issue inviting someone with CUDA knowledge to further improve upon your work.

This is just a suggestion, of course.

File: megatron/model/transformer.py (hunk 2 above)
Comment (ofirpress) [reply megatron/model/transformer.py:688]: > does it break if don't to() explicity, @ofirpress?
Yes it says that the alibi tensor is on the CPU if I don't to() it.

>The other approach to not needing to use torch.cuda.current_device() in forward is to use .device of one of the inputs.
But during init do we have any of the inputs yet?

File: megatron/model/transformer.py (hunk 3 above)
Comment (ofirpress) [reply megatron/model/transformer.py]: Thank you, I've now changed it to .to(torch.bfloat16)

Comment (stas00) [reply megatron/model/transformer.py]: no, no. it was correct in the first place.

@ibeltagy  just suggested to also support `bf16` and therefore to add `elif args.params_dtype == torch.bfloat16: self.alibi = self.alibi.to(torch.bfloat16)`

but for symmetry, probably let's do the first branch `self.alibi = self.alibi.to(torch.float16)`

In other words:

```
            if args.params_dtype == torch.float16:
                self.alibi = self.alibi.to(torch.float16)
            elif args.params_dtype == torch.bfloat16:
                self.alibi = self.alibi.to(torch.bfloat16)

```

File: megatron/model/transformer.py (hunk 2 above)
Comment (stas00) [reply megatron/model/transformer.py:688]: In init you typically have other params, but it's probably OK the way you did it. 

Usually the model is made of registered params and buffers and those get automatically switched to the right device with the sub-module, so custom tensors not attached to the model are tricky. and typically you have to update them in `forward` to be on the same device as inputs or params.

I haven't looked at the full context, so let's revisit this when the tests are added if you're running into problems.

File: megatron/model/transformer.py (hunk 3 above)
Comment (ofirpress) [reply megatron/model/transformer.py]: Ah sorry, that makes sense. Fixed.

File: megatron/model/transformer.py (hunk 2 above)
Comment (ofirpress) [reply megatron/model/transformer.py:688]: Ya we can change it in the forward if this approach doesn't work.

Comment (stas00) [reply megatron/model/transformer.py:688]: If it can be done in forward, then that's where we want it done I believe. Because only in `forward` you know which device you're on. You can't rely on the where it was init'ed. If it makes sense.

Comment (stas00) [reply megatron/model/transformer.py:688]: for example:
https://github.com/huggingface/transformers/blob/master/src/transformers/models/fsmt/modeling_fsmt.py#L1276

Comment (ofirpress) [reply megatron/model/transformer.py:688]: I guess I'm not sure what the elegant way to do it is. 
I guess we can do something like:
if self.alibi.device != some_input.device
    self.alibi.to(some_input.device)

but I'm not sure if that's OK.

Comment (stas00) [reply megatron/model/transformer.py:688]: there is no need for if. It's a noop if it's already on the right device. So just the to call (plus assignment) as in the example I linked to.

File: megatron/model/transformer.py [hunk 4]
```diff
@@ -594,6 +597,27 @@ def forward(self, inputs, **kwargs):
...
+            def get_slopes_power_of_2(n):
+                start = (2 ** (-2 ** -(math.log2(n) - 3)))
+                ratio = start
+                return [start * ratio ** i for i in range(n)]
+
+            if math.log2(n).is_integer():
+                return get_slopes_power_of_2(n)
+            else:
+                closest_power_of_2 = 2 ** math.floor(math.log2(n))
+                return get_slopes_power_of_2(closest_power_of_2) + get_slopes(2 * closest_power_of_2)[0::2][
```
Comment (thomasw21) [inline megatron/model/transformer.py:614]: Why is it important to check that `n` is always a power of 2?

File: megatron/model/transformer.py [hunk 5]
```diff
@@ -594,6 +597,27 @@ def forward(self, inputs, **kwargs):
...
+                                                                   :n - closest_power_of_2]
+        slopes = torch.Tensor(get_slopes(num_attention_heads))
+        alibi = slopes.unsqueeze(1).unsqueeze(1) * torch.arange(max_seq_len).unsqueeze(0).unsqueeze(0).expand(num_attention_heads, -1, -1)
```
Comment (thomasw21) [inline megatron/model/transformer.py:617]:
```suggestion
        alibi = slopes[:, None, None] * torch.arange(max_seq_len)[None, None, :].expand(num_attention_heads, -1, -1)
```

Also can you import the same comment as your original implementation? Typically where the matrix here doesn't match the one on the paper?

File: megatron/model/transformer.py [hunk 6]
```diff
@@ -594,6 +597,27 @@ def forward(self, inputs, **kwargs):
...
+        alibi = alibi.view(num_attention_heads, 1, max_seq_len).repeat(batch_size, 1, 1)
```
Comment (thomasw21) [inline megatron/model/transformer.py:618]: Isn't alibi already the correct format? Also why not expand?

```suggestion
        alibi = alibi.repeat(batch_size, 1, 1)
```

File: megatron/model/transformer.py [hunk 7]
```diff
@@ -277,12 +277,15 @@ def forward(self, hidden_states, attention_mask, layer_past=None,
...
-            device=torch.cuda.current_device())
+        if alibi is None:
+            matmul_result = torch.empty(
+                output_size[0]*output_size[1],
+                output_size[2],
+                output_size[3],
+                dtype=query_layer.dtype,
+                device=torch.cuda.current_device())
+        else:
+            matmul_result = alibi[:output_size[0]*output_size[1], :, :output_size[3]]
```
Comment (thomasw21) [inline megatron/model/transformer.py:288]: https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/101/files#diff-46c4c76deb18adf1de8e0be6d4229baed5f1f0308e141479f5a993b3d83dd445R307

When using `baddmm` shouldn't you set beta to 1? otherwise alibi should be ignored no?
https://pytorch.org/docs/stable/generated/torch.baddbmm.html

If you end up modifying `beta` then you probably have to replace `empty` with `zeros` in `if alibi is None`

**Review by thomasw21 — COMMENTED** [review]
Thanks ! I have some questions concerning current implementation:
 - We usually have a `--reset-position-ids` to reset the position ids within a row when we hit of end of document token, ie in a row we could have [0,1,2,3,4,0,1,2,3] (in the absolute embedding implementation). This does mean that it will conflict with the notion of static alibi matrix though. I don't think current implementation handles this case?
  - We have two options `--reset-position-ids` and `--reset-attention-mask` which up until now were quite distinct, but it starts to get mixed up with alibi. I'm not sure what the behaviour of alibi should be for cases where one is True the other one is False ...

**Comment by ibeltagy** [comment]
>  `--reset-position-ids` and `--reset-attention-mask`

We do not need to worry about this right now. We can do something like `if alibi and (reset position id or resent attention mask) raise error`

File: megatron/model/transformer.py (hunk 4 above)
Comment (ibeltagy) [reply megatron/model/transformer.py:614]: I think this is discussed in the paper

File: megatron/model/transformer.py (hunk 2 above)
Comment (ibeltagy) [reply megatron/model/transformer.py:688]: I am happy merging the PR as is then fix this later if it ends up being a problem. As @stas00, if it is wrong, pytorch will complain.

File: megatron/model/transformer.py (hunk 7 above)
Comment (ibeltagy) [reply megatron/model/transformer.py:288]: I see, you are right. beta here https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/3350322698f6332d3b4eb993c08145e11578a31e/megatron/model/transformer.py#L307 should be 1.0 instead

Comment (thomasw21) [reply megatron/model/transformer.py:288]: How does the paper handle the normalizing factor? Is it applied after the sum?

File: megatron/model/transformer.py (hunk 4 above)
Comment (ofirpress) [reply megatron/model/transformer.py:614]: Explained in the comment here: https://github.com/ofirpress/attention_with_linear_biases/blob/master/fairseq/models/transformer.py#L749

```
#In the paper, we only train models that have 2^a heads for some a. This function has
#some good properties that only occur when the input is a power of 2. To maintain that even
#when the number of heads is not a power of 2, we use this workaround. 
```

File: megatron/model/transformer.py (hunk 7 above)
Comment (ofirpress) [reply megatron/model/transformer.py:288]: > When using baddmm shouldn't you set beta to 1? 

Yes!!! I meant to do that and then totally forgot! Thanks so much for pointing this out!!!

> If you end up modifying beta then you probably have to replace empty with zeros in if alibi is None

Wouldn't it be better to do `beta = 1 if alibi is not None else 0`?

>  How does the paper handle the normalizing factor? Is it applied after the sum?

Which normalizing factor are you referring to? The softmax? If so, we apply the softmax after adding the ALiBi bias.

Comment (thomasw21) [reply megatron/model/transformer.py:288]: I mean the `1.0 / self.norm_factor` part. Is it applied only on `QK` or on the entire attention matrix?

Comment (ofirpress) [reply megatron/model/transformer.py:288]: Ah ok, I understand. We apply it before the sum, see https://github.com/ofirpress/attention_with_linear_biases/blob/master/fairseq/modules/multihead_attention.py#L225

**Comment by ofirpress** [comment]
> Thanks ! I have some questions concerning current implementation:
> 
> * We usually have a `--reset-position-ids` to reset the position ids within a row when we hit of end of document token, ie in a row we could have [0,1,2,3,4,0,1,2,3] (in the absolute embedding implementation). This does mean that it will conflict with the notion of static alibi matrix though. I don't think current implementation handles this case?
> * We have two options `--reset-position-ids` and `--reset-attention-mask` which up until now were quite distinct, but it starts to get mixed up with alibi. I'm not sure what the behaviour of alibi should be for cases where one is True the other one is False ...

I'm not quite sure why you would want to reset the absolute position embeddings when starting a new document, but with relative position embeddings there is no need for that I'm pretty sure. If you have an attention mask that masks out other documents from the current one, then that will directly also work with ALiBi and any other relative position method, no modifications needed there.

File: megatron/model/transformer.py (hunk 6 above)
Comment (ofirpress) [reply megatron/model/transformer.py]: Ya that view is not necessary, thanks!

About repeat vs expand-
Expand can only be used to expand dims that are of size 1, so it can't be used here unfortunately :(

**Comment by stas00** [comment]
A quick note to reviewers - please make sure that all new features include extensive multi-gpu testing before the feature is merged. 

You will find the existing tests to build on under `tests/`, that you can use to build upon and if you need help with anything testing-wise or just want to validate the test looks good please don't hesitate to tag me, thank you!

File: megatron/model/transformer.py (hunk 7 above)
Comment (ofirpress) [reply megatron/model/transformer.py:288]: Fixed now :)

**Comment by ofirpress** [comment]
I just fixed the bug with the beta value in the attention computation. I've tested this by running the pretrain_gpt example twice, one with ALiBi and once without, and they both train until the end, with the ALiBi model achieving better performance which means its actually working (unlike before where the ALiBi matrix was being ignored). 

Thanks so much @thomasw21 for pointing out that bug!

**Comment by thomasw21** [comment]
Hum smells like a bad merge. Any chance you merged with `master` instead of `main`?

**Comment by ofirpress** [comment]
I'm not great with git so that's definitely possible. What can I do to fix this? Thanks!

**Comment by thomasw21** [comment]
Can you allow me to write on your fork? I can try to fix it.

**Comment by SaulLu** [comment]
It looks like the last change you made to this PR is the `bugfix with beta value in attention computation` commit with the `
5b5afb2 ` hash.

If this is the case, what you should do is undo revert all subsequent commits. There are two ways to do this:
1. by rewriting history
2. by creating commits that will undo one by one the commits you want to remove.

To use the second method, note that the first and last hash commits to be cancelled are `22a6e17` and `a49461c`.

It seems to me that the appropriate command would be:
```
git revert 22a6e17...a49461c
```

**Comment by thomasw21** [comment]
This looks about right, it's just that you're going to need to force push which is always a bit risky.

**Comment by SaulLu** [comment]
@thomasw21, if the second option is chosen, it seems to me that there will be no need for a force push (which is perhaps the most "reassuring" thing to do) :slightly_smiling_face:

**Comment by ofirpress** [comment]
Thanks so much @SaulLu and @thomasw21!

**Comment by thomasw21** [comment]
Can we merge this @SaulLu ?

**Review by TevenLeScao — APPROVED** [review]
Let's merge, the main thing is this `to(device)` that we needed to check, but it seems to have worked so far.
